File: scripts/solar/solar-common/actor/ActorUnit.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ClassExtends = ____lualib.__TS__ClassExtends
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local __TS__New = ____lualib.__TS__New
local __TS__SetDescriptor = ____lualib.__TS__SetDescriptor
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["10"] = 1,["11"] = 1,["12"] = 2,["13"] = 2,["14"] = 3,["15"] = 3,["16"] = 4,["17"] = 4,["18"] = 5,["19"] = 5,["20"] = 7,["21"] = 7,["22"] = 14,["23"] = 14,["24"] = 14,["25"] = 14,["26"] = 21,["27"] = 14,["28"] = 19,["29"] = 24,["30"] = 25,["31"] = 26,["32"] = 28,["34"] = 30,["35"] = 32,["37"] = 34,["38"] = 35,["41"] = 39,["42"] = 40,["44"] = 43,["45"] = 43,["46"] = 43,["47"] = 43,["48"] = 43,["49"] = 43,["50"] = 43,["51"] = 45,["52"] = 46,["53"] = 47,["54"] = 49,["56"] = 50,["57"] = 50,["58"] = 51,["59"] = 53,["60"] = 54,["61"] = 54,["62"] = 54,["63"] = 54,["66"] = 50,["70"] = 21,["71"] = 68,["72"] = 14,["73"] = 70,["77"] = 73,["78"] = 154,["79"] = 74,["81"] = 75,["82"] = 75,["84"] = 76,["85"] = 76,["89"] = 78,["91"] = 79,["92"] = 80,["94"] = 82,["95"] = 84,["96"] = 84,["101"] = 87,["103"] = 88,["106"] = 90,["108"] = 91,["109"] = 91,["110"] = 91,["111"] = 91,["112"] = 91,["113"] = 91,["114"] = 91,["117"] = 94,["119"] = 95,["120"] = 95,["121"] = 95,["122"] = 95,["123"] = 95,["124"] = 95,["125"] = 95,["128"] = 98,["130"] = 99,["131"] = 99,["132"] = 99,["133"] = 99,["134"] = 99,["135"] = 99,["136"] = 99,["139"] = 102,["141"] = 103,["142"] = 103,["143"] = 103,["144"] = 103,["145"] = 103,["146"] = 103,["147"] = 103,["150"] = 106,["152"] = 107,["153"] = 107,["157"] = 109,["159"] = 110,["162"] = 112,["164"] = 113,["165"] = 113,["166"] = 113,["167"] = 113,["168"] = 114,["169"] = 114,["170"] = 114,["171"] = 114,["172"] = 115,["175"] = 117,["177"] = 118,["178"] = 119,["181"] = 121,["183"] = 122,["184"] = 123,["187"] = 126,["189"] = 127,["192"] = 129,["194"] = 130,["197"] = 132,["199"] = 133,["202"] = 135,["204"] = 136,["207"] = 138,["209"] = 139,["212"] = 141,["214"] = 142,["217"] = 144,["219"] = 145,["222"] = 147,["224"] = 148,["227"] = 150,["229"] = 151,["232"] = 153,["234"] = 154,["235"] = 155,["236"] = 156,["240"] = 159,["242"] = 160,["243"] = 161,["244"] = 162,["249"] = 68,["250"] = 169,["251"] = 14,["252"] = 169,["253"] = 173,["254"] = 14,["255"] = 175,["256"] = 173,["257"] = 178,["258"] = 178,["259"] = 178,["261"] = 179,["262"] = 180,["264"] = 14,["265"] = 178,["266"] = 188,["267"] = 188,["268"] = 189,["271"] = 192,["272"] = 193,["275"] = 197,["276"] = 198,["277"] = 199,["278"] = 200,["279"] = 201,["280"] = 201,["283"] = 188,["284"] = 210,["285"] = 211,["286"] = 212,["287"] = 213,["288"] = 214,["291"] = 217,["292"] = 210,["293"] = 224,["294"] = 225,["296"] = 226,["297"] = 226,["298"] = 227,["299"] = 228,["300"] = 229,["301"] = 230,["303"] = 226,["306"] = 233,["307"] = 224,["308"] = 236,["309"] = 236,["310"] = 236,["312"] = 237,["313"] = 236,["314"] = 244,["315"] = 244,["316"] = 245,["317"] = 246,["318"] = 247,["319"] = 248,["320"] = 249,["323"] = 252,["327"] = 244,["328"] = 264,["329"] = 265,["330"] = 266,["331"] = 264,["332"] = 274,["333"] = 275,["334"] = 275,["335"] = 276,["338"] = 279,["339"] = 280,["340"] = 280,["341"] = 281,["342"] = 282,["345"] = 285,["346"] = 274,["347"] = 293,["348"] = 294,["349"] = 294,["350"] = 295,["351"] = 296,["353"] = 298,["354"] = 299,["355"] = 299,["356"] = 300,["357"] = 301,["360"] = 293,["361"] = 306,["362"] = 307,["363"] = 14,["364"] = 309,["365"] = 310,["367"] = 313,["368"] = 314,["370"] = 306,["371"] = 16,["376"] = 65});
local ____exports = {}
local ____Actor = require("solar.solar-common.actor.Actor")
local Actor = ____Actor.default
local ____ObjectTemplateUtil = require("solar.solar-common.util.object.ObjectTemplateUtil")
local ObjectTemplateUtil = ____ObjectTemplateUtil.default
local ____DataBase = require("solar.solar-common.common.DataBase")
local DataBase = ____DataBase.default
local ____ActorAbility = require("solar.solar-common.actor.ActorAbility")
local ActorAbility = ____ActorAbility.default
local ____ActorTypeShopUtil = require("solar.solar-common.actor.util.ActorTypeShopUtil")
local ActorTypeShopUtil = ____ActorTypeShopUtil.default
local ____UnitStateUtil = require("solar.solar-common.util.unit.UnitStateUtil")
local UnitStateUtil = ____UnitStateUtil.default
____exports.default = __TS__Class()
local ActorUnit = ____exports.default
ActorUnit.name = "ActorUnit"
__TS__ClassExtends(ActorUnit, Actor)
function ActorUnit.prototype.____constructor(self, actorTypeId, player, x, y)
    Actor.prototype.____constructor(self, actorTypeId)
    self.sellItemAbilityList = {}
    local templateType = self:getTemplateType()
    ObjectTemplateUtil:_sl_init()
    if ObjectTemplateUtil.templateInfo[templateType] == nil then
        self.unitTypeID = templateType
    else
        self.unitTypeID = ObjectTemplateUtil:borrowTemplate(templateType)
        ObjectTemplateUtil:returnTemplate(templateType, self.unitTypeID)
    end
    if self.unitTypeID == nil or #self.unitTypeID ~= 4 then
        log.errorWithTraceBack((("物编模板id获取失败:" .. tostring(templateType)) .. " -> ") .. tostring(self.unitTypeID))
        return
    end
    if self.actorType.sellItems and #self.actorType.sellItems > 0 and self.actorType.onUnitInterval == nil then
        ActorTypeShopUtil:autoDisableSellingAbilityOnNoStoreTarget(actorTypeId)
    end
    self.unit = CreateUnit(
        player,
        self.unitTypeID,
        x,
        y,
        270
    )
    DataBase:getUnitSolarData(self.unit)._SL_solarActorUnit = self
    self:_sl_init()
    self:update()
    if IsHeroUnitId(GetUnitTypeId(self.unit)) and FrameSetOriginButtonTexture and GetLocalPlayer() == player then
        do
            local i = 5
            while i >= 0 do
                local heroBarButton = DzFrameGetHeroBarButton(i)
                if FrameIsShow(heroBarButton) then
                    FrameSetOriginButtonTexture(
                        heroBarButton,
                        self:getIcon()
                    )
                    break
                end
                i = i - 1
            end
        end
    end
end
function ActorUnit.prototype._sl_rawset(self, key, value)
    Actor.prototype._sl_rawset(self, key, value)
    if not IsHandle(self.unit) then
        return
    end
    repeat
        local ____switch12 = key
        local sellItems, abilities
        local ____cond12 = ____switch12 == "name"
        if ____cond12 then
            if SetUnitName ~= nil then
                SetUnitName(self.unit, value)
            end
            if SetUnitProperName ~= nil then
                SetUnitProperName(self.unit, value)
            end
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "model"
        if ____cond12 then
            if SetUnitModel then
                SetUnitModel(self.unit, value)
            else
                DzSetUnitModel(self.unit, value)
                if SetUnitPortrait ~= nil then
                    SetUnitPortrait(self.unit, value)
                end
            end
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "modelScale"
        if ____cond12 then
            SetUnitScale(self.unit, value, value, value)
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "colorR"
        if ____cond12 then
            SetUnitVertexColor(
                self.unit,
                value or 255,
                self:get("colorG", 255),
                self:get("colorB", 255),
                self:get("alpha", 255)
            )
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "colorG"
        if ____cond12 then
            SetUnitVertexColor(
                self.unit,
                self:get("colorR", 255),
                value or 255,
                self:get("colorB", 255),
                self:get("alpha", 255)
            )
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "colorB"
        if ____cond12 then
            SetUnitVertexColor(
                self.unit,
                self:get("colorR", 255),
                self:get("colorG", 255),
                value or 255,
                self:get("alpha", 255)
            )
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "alpha"
        if ____cond12 then
            SetUnitVertexColor(
                self.unit,
                self:get("colorR", 255),
                self:get("colorG", 255),
                self:get("colorB", 255),
                value or 255
            )
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "missileModel"
        if ____cond12 then
            if SetUnitMissileModel ~= nil then
                SetUnitMissileModel(self.unit, value)
            end
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "moveType"
        if ____cond12 then
            UnitStateUtil:setMoveType(self.unit, value)
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "flyHeight"
        if ____cond12 then
            UnitAddAbility(
                self.unit,
                FourCC("Amrf")
            )
            UnitRemoveAbility(
                self.unit,
                FourCC("Amrf")
            )
            SetUnitFlyHeight(self.unit, value, 0)
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "maxLife"
        if ____cond12 then
            SetUnitState(self.unit, UNIT_STATE_MAX_LIFE, value)
            SetUnitState(self.unit, UNIT_STATE_LIFE, value)
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "maxMana"
        if ____cond12 then
            SetUnitState(self.unit, UNIT_STATE_MAX_MANA, value)
            SetUnitState(self.unit, UNIT_STATE_MANA, value)
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "acquireRange"
        if ____cond12 then
            SetUnitAcquireRange(self.unit, value)
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "range"
        if ____cond12 then
            SetUnitState(self.unit, UnitStateDamageRange, value)
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "damage"
        if ____cond12 then
            SetUnitState(self.unit, UnitStateDamageBase, value)
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "damageCd"
        if ____cond12 then
            SetUnitState(self.unit, UnitStateDamageCool, value)
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "strength"
        if ____cond12 then
            SetHeroStr(self.unit, value, true)
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "agility"
        if ____cond12 then
            SetHeroAgi(self.unit, value, true)
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "intelligence"
        if ____cond12 then
            SetHeroInt(self.unit, value, true)
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "def"
        if ____cond12 then
            SetUnitState(self.unit, UnitStateArmor, value)
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "moveSpeed"
        if ____cond12 then
            SetUnitMoveSpeed(self.unit, value)
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "sellItems"
        if ____cond12 then
            sellItems = value
            if sellItems and #sellItems > 0 then
                self:addSellItem(table.unpack(sellItems))
            end
            break
        end
        ____cond12 = ____cond12 or ____switch12 == "abilities"
        if ____cond12 then
            abilities = value
            if abilities and #abilities > 0 then
                self:addAbility(table.unpack(abilities))
            end
            break
        end
    until true
end
function ActorUnit.prototype.get(self, key, defaultValue)
    return Actor.prototype.get(self, key, defaultValue)
end
function ActorUnit.prototype.setXY(self, x, y)
    Actor.prototype.setXY(self, x, y)
    SetUnitPosition(self.unit, x, y)
end
function ActorUnit.prototype.getRootFrameControl(self, createDefault)
    if createDefault == nil then
        createDefault = true
    end
    if createDefault and self._sl_rootFrameControl == nil then
        ____exports.default._sl_hasFrameActorUnits[self.uuid] = self
    end
    return Actor.prototype.getRootFrameControl(self, createDefault)
end
function ActorUnit.prototype.addSellItem(self, ...)
    local sellItems = {...}
    if sellItems == nil or #sellItems == 0 then
        return
    end
    if #sellItems > 12 then
        log.errorWithTraceBack((self:getName() .. ":sellItems售卖物品超过12个。") .. tostring(#sellItems))
        return
    end
    for ____, sellItemId in ipairs(sellItems) do
        if not self:hasSellItem(sellItemId) then
            local sellingAbilityTypeId = ActorTypeShopUtil:warpActorItem2SellingAbility(sellItemId).id
            local actorAbility = self:addActorAbility(sellingAbilityTypeId)
            local ____self_sellItemAbilityList_8 = self.sellItemAbilityList
            ____self_sellItemAbilityList_8[#____self_sellItemAbilityList_8 + 1] = actorAbility
        end
    end
end
function ActorUnit.prototype.hasSellItem(self, sellItemId)
    local sellingAbilityTypeId = ActorTypeShopUtil:warpActorItem2SellingAbility(sellItemId).id
    for ____, sellItemAbility in ipairs(self.sellItemAbilityList) do
        if sellItemAbility.actorType.id == sellingAbilityTypeId then
            return true
        end
    end
    return false
end
function ActorUnit.prototype.removeSellItem(self, sellItemId)
    local sellingAbilityTypeId = ActorTypeShopUtil:warpActorItem2SellingAbility(sellItemId).id
    do
        local i = 0
        while i < #self.sellItemAbilityList do
            local sellItemAbility = self.sellItemAbilityList[i + 1]
            if sellItemAbility.actorType.id == sellingAbilityTypeId then
                sellItemAbility:destroy()
                self.sellItemAbilityList = __TS__ArraySplice(self.sellItemAbilityList, i, 1)
            end
            i = i + 1
        end
    end
    return false
end
function ActorUnit.prototype.applyTimedLife(self, duration, buffid)
    if buffid == nil then
        buffid = "BHwe"
    end
    UnitApplyTimedLife(self.unit, buffid, duration)
end
function ActorUnit.prototype.addAbility(self, ...)
    local abilities = {...}
    if abilities and #abilities > 0 then
        for ____, ability in ipairs(abilities) do
            if DataBase:getSolarActorType(ability) ~= nil then
                if self:getActorAbility(ability) == nil then
                    self:addActorAbility(ability)
                end
            else
                UnitAddAbility(self.unit, ability)
            end
        end
    end
end
function ActorUnit.prototype.addActorAbility(self, actorAbilityTypeId, posNum)
    local actorAbility = __TS__New(ActorAbility, actorAbilityTypeId, self.unit, posNum)
    return actorAbility
end
function ActorUnit.prototype.getActorAbility(self, actorTypeId)
    local ____opt_9 = DataBase:getUnitSolarData(self.unit, false)
    local abilityTemplate = ____opt_9 and ____opt_9._SL_abilityTemplate
    if abilityTemplate == nil then
        return
    end
    for abilityTemplateKey in pairs(abilityTemplate) do
        local ____opt_11 = DataBase:getAbilityTypeSolarData(abilityTemplate[abilityTemplateKey], false)
        local actor = ____opt_11 and ____opt_11._SL_solarActorAbility
        if actor ~= nil and actor.actorTypeId == actorTypeId then
            return actor
        end
    end
    return nil
end
function ActorUnit.prototype.destroyAllActorAbility(self)
    local ____opt_13 = DataBase:getUnitSolarData(self.unit, false)
    local abilityTemplate = ____opt_13 and ____opt_13._SL_abilityTemplate
    if abilityTemplate == nil then
        return nil
    end
    for abilityTemplateKey in pairs(abilityTemplate) do
        local ____opt_15 = DataBase:getAbilityTypeSolarData(abilityTemplate[abilityTemplateKey], false)
        local actor = ____opt_15 and ____opt_15._SL_solarActorAbility
        if actor ~= nil then
            actor:destroy()
        end
    end
end
function ActorUnit.prototype.destroy(self)
    local unit = self.unit
    Actor.prototype.destroy(self)
    if self._sl_RootFrame ~= nil then
        deleteKey(____exports.default._sl_hasFrameActorUnits, self.uuid)
    end
    if IsHandle(unit) then
        RemoveUnit(unit)
    end
end
ActorUnit._sl_hasFrameActorUnits = {}
__TS__SetDescriptor(
    ActorUnit.prototype,
    "actorType",
    {get = function(self)
        return self._actorType
    end},
    true
)
return ____exports

File: src/solar/solar-common/actor/ActorUnit.ts
import Actor from "@/Actor";
import ObjectTemplateUtil from "@/ObjectTemplateUtil";
import DataBase from "@/DataBase";
import ActorAbility from "@/ActorAbility";
import ActorTypeShopUtil from "@/ActorTypeShopUtil";
import FrameControl from "@/FrameControl";
import UnitStateUtil from "@/UnitStateUtil";

/**
 * 演员单位
 * 演员单位的类型是公用的。
 *
 */
export default class ActorUnit extends Actor {
    /** 拥有UI的单位 比如模拟血条*/
    static _sl_hasFrameActorUnits: { [uuid: string]: ActorUnit } = {}
    //
    unitTypeID: string
    sellItemAbilityList: ActorAbility[] = []

    constructor(actorTypeId: string, player: player, x: number, y: number) {
        super(actorTypeId)
        //单位模板 即用即还
        let templateType = this.getTemplateType();
        ObjectTemplateUtil._sl_init();
        if (ObjectTemplateUtil.templateInfo[templateType] == null) {
            //直接使用原生物编id 当模板
            this.unitTypeID = templateType;
        } else {
            this.unitTypeID = ObjectTemplateUtil.borrowTemplate(templateType);
            //单位模板 即用即还
            ObjectTemplateUtil.returnTemplate(templateType, this.unitTypeID);
        }
        if (this.unitTypeID == null || this.unitTypeID.length != 4) {
            log.errorWithTraceBack("物编模板id获取失败:" + tostring(templateType) + " -> " + tostring(this.unitTypeID))
            return;
        }
        //
        if (this.actorType.sellItems && this.actorType.sellItems.length > 0 && this.actorType.onUnitInterval == null) {
            ActorTypeShopUtil.autoDisableSellingAbilityOnNoStoreTarget(actorTypeId)
        }
        //
        this.unit = CreateUnit(player, this.unitTypeID, x, y, 270)
        //
        DataBase.getUnitSolarData(this.unit)._SL_solarActorUnit = this;
        this._sl_init();
        this.update();
        //设置英雄左上角图标
        if (IsHeroUnitId(GetUnitTypeId(this.unit)) && FrameSetOriginButtonTexture && GetLocalPlayer() == player) {
            for (let i = 5; i >= 0; i--) {
                let heroBarButton = DzFrameGetHeroBarButton(i);
                //把最后一个图标 设置为正确的英雄图标
                if (FrameIsShow(heroBarButton)) {
                    FrameSetOriginButtonTexture(heroBarButton, this.getIcon());
                    break;
                }
            }
        }
    }

    /**
     * 获取演员的类型
     */
    get actorType(): AppActorUnitType {
        return this._actorType;
    }

    protected _sl_rawset<Key extends keyof AppActorUnitType>(key: Key, value: AppActorUnitType[Key]) {
        super._sl_rawset(key, value);
        if (!IsHandle(this.unit)) {
            return
        }
        switch (key) {
            case "name":
                SetUnitName?.(this.unit, value)
                SetUnitProperName?.(this.unit, value)
                break;
            case "model":
                if (SetUnitModel) {
                    SetUnitModel(this.unit, value)
                } else {
                    DzSetUnitModel(this.unit, value)
                    //设置大头像
                    SetUnitPortrait?.(this.unit, value)
                }
                break;
            case "modelScale":
                SetUnitScale(this.unit, value, value, value)
                break;
            case "colorR":
                SetUnitVertexColor(this.unit, value || 255, this.get("colorG", 255),
                    this.get("colorB", 255), this.get("alpha", 255))
                break;
            case "colorG":
                SetUnitVertexColor(this.unit, this.get("colorR", 255), value || 255,
                    this.get("colorB", 255), this.get("alpha", 255))
                break;
            case "colorB":
                SetUnitVertexColor(this.unit, this.get("colorR", 255), this.get("colorG", 255),
                    value || 255, this.get("alpha", 255))
                break;
            case "alpha":
                SetUnitVertexColor(this.unit, this.get("colorR", 255), this.get("colorG", 255),
                    this.get("colorB", 255), value || 255)
                break;
            case "missileModel":
                SetUnitMissileModel?.(this.unit, value)
                break;
            case "moveType":
                UnitStateUtil.setMoveType(this.unit, value)
                break;
            case "flyHeight":
                UnitAddAbility(this.unit, FourCC('Amrf'));
                UnitRemoveAbility(this.unit, FourCC('Amrf'));
                SetUnitFlyHeight(this.unit, value, 0)
                break;
            case "maxLife":
                SetUnitState(this.unit, UNIT_STATE_MAX_LIFE, value)
                SetUnitState(this.unit, UNIT_STATE_LIFE, value)
                break;
            case "maxMana":
                SetUnitState(this.unit, UNIT_STATE_MAX_MANA, value)
                SetUnitState(this.unit, UNIT_STATE_MANA, value)
                break;
            //设置感知敌人距离
            case "acquireRange":
                SetUnitAcquireRange(this.unit, value)
                break;
            case "range":
                SetUnitState(this.unit, UnitStateDamageRange, value)
                break;
            case "damage":
                SetUnitState(this.unit, UnitStateDamageBase, value)
                break;
            case "damageCd":
                SetUnitState(this.unit, UnitStateDamageCool, value)
                break;
            case "strength":
                SetHeroStr(this.unit, value, true)
                break;
            case "agility":
                SetHeroAgi(this.unit, value, true)
                break;
            case "intelligence":
                SetHeroInt(this.unit, value, true)
                break;
            case "def":
                SetUnitState(this.unit, UnitStateArmor, value)
                break;
            case "moveSpeed":
                SetUnitMoveSpeed(this.unit, value)
                break;
            case "sellItems":
                let sellItems: string[] = value as any;
                if (sellItems && sellItems.length > 0) {
                    this.addSellItem(...sellItems)
                }
                break;
            case "abilities":
                let abilities: string[] = value as any;
                if (abilities && abilities.length > 0) {
                    this.addAbility(...abilities)
                }
                break;
        }
    }


    get<Key extends keyof AppActorUnitType>(key: Key, defaultValue?: AppActorUnitType[Key]): AppActorUnitType[Key] {
        return super.get(key, defaultValue);
    }

    setXY(x: number, y: number) {
        super.setXY(x, y);
        SetUnitPosition(this.unit, x, y)
    }

    getRootFrameControl(createDefault: boolean = true): FrameControl {
        if (createDefault && this._sl_rootFrameControl == null) {
            ActorUnit._sl_hasFrameActorUnits[this.uuid] = this;
        }
        return super.getRootFrameControl(createDefault);
    }

    /**
     * 安全的添加售卖物品。区别是可以重复调用此方法 不会添加重复的物品去售卖
     */
    addSellItem(...sellItems: string[]) {
        if (sellItems == null || sellItems.length == 0) {
            return
        }
        if (sellItems.length > 12) {
            log.errorWithTraceBack(this.getName() + ":sellItems售卖物品超过12个。" + sellItems.length)
            return;
        }

        for (let sellItemId of sellItems) {
            if (!this.hasSellItem(sellItemId)) {
                let sellingAbilityTypeId = ActorTypeShopUtil.warpActorItem2SellingAbility(sellItemId).id;
                let actorAbility = this.addActorAbility(sellingAbilityTypeId);
                this.sellItemAbilityList.push(actorAbility);
            }
        }
    }

    /**
     * 是否已有某个演员物品的售卖
     * @param sellItemId
     */
    hasSellItem(sellItemId: string): boolean {
        let sellingAbilityTypeId = ActorTypeShopUtil.warpActorItem2SellingAbility(sellItemId).id;
        for (let sellItemAbility of this.sellItemAbilityList) {
            if (sellItemAbility.actorType.id == sellingAbilityTypeId) {
                return true;
            }
        }
        return false;
    }

    /**
     * 移除某个演员物品的售卖
     * @param sellItemId
     */
    removeSellItem(sellItemId: string): boolean {
        let sellingAbilityTypeId = ActorTypeShopUtil.warpActorItem2SellingAbility(sellItemId).id;
        for (let i = 0; i < this.sellItemAbilityList.length; i++) {
            let sellItemAbility = this.sellItemAbilityList[i];
            if (sellItemAbility.actorType.id == sellingAbilityTypeId) {
                sellItemAbility.destroy();
                this.sellItemAbilityList = this.sellItemAbilityList.splice(i, 1)
            }
        }
        return false;
    }

    applyTimedLife(duration: number, buffid: number | string = 'BHwe') {
        UnitApplyTimedLife(this.unit, buffid, duration)
    }


    /**
     * 添加技能(或者演员技能)
     */
    addAbility(...abilities: string[]): void {
        if (abilities && abilities.length > 0) {
            for (let ability of abilities) {
                if (DataBase.getSolarActorType(ability) != null) {
                    if (this.getActorAbility(ability) == null) {
                        this.addActorAbility(ability)
                    }
                } else {
                    UnitAddAbility(this.unit, ability)
                }
            }
        }
    }

    /**
     * 添加一个演员技能
     * @param actorAbilityTypeId
     * @param posNum 指定技能位置
     * @param startPos 若为指定技能位置则从技能起始位置开始找到后面的空位值
     */
    addActorAbility(actorAbilityTypeId: string, posNum?: number): ActorAbility {
        let actorAbility = new ActorAbility(actorAbilityTypeId, this.unit, posNum);
        return actorAbility;
    }


    /**
     * 获取单位身上的指定类型的演员技能
     * @param actorTypeId
     */
    getActorAbility(actorTypeId: string): ActorAbility {
        let abilityTemplate = DataBase.getUnitSolarData(this.unit, false)?._SL_abilityTemplate;
        if (abilityTemplate == null) {
            return;
        }
        for (let abilityTemplateKey in abilityTemplate) {
            let actor = DataBase.getAbilityTypeSolarData(abilityTemplate[abilityTemplateKey], false)?._SL_solarActorAbility;
            if (actor != null && actor.actorTypeId == actorTypeId) {
                return actor as ActorAbility;
            }
        }
        return null;
    }


    /**
     * 销毁此单位身上的所有 演员技能
     * @param unit
     */
    destroyAllActorAbility(): void {
        let abilityTemplate = DataBase.getUnitSolarData(this.unit, false)?._SL_abilityTemplate;
        if (abilityTemplate == null) {
            return null;
        }
        for (let abilityTemplateKey in abilityTemplate) {
            let actor = DataBase.getAbilityTypeSolarData(abilityTemplate[abilityTemplateKey], false)?._SL_solarActorAbility;
            if (actor != null) {
                actor.destroy();
            }
        }
    }

    destroy() {
        let unit = this.unit
        super.destroy();
        if (this._sl_RootFrame != null) {
            deleteKey(ActorUnit._sl_hasFrameActorUnits, this.uuid)
        }
        //这里可能为null 比如多次调用了destroy 后面调用的都没有东西可以清理了
        if (IsHandle(unit)) {
            RemoveUnit(unit)
        }
    }

}